language/move-vm/runtime/src/native_functions.rs

```diff
@@ -172,4 +176,35 @@ impl<'a, 'b> NativeContext<'a, 'b> {
     pub fn stack_frames(&self, count: usize) -> ExecutionState {
         self.interpreter.get_stack_frames(count)
     }
+
+    pub fn gas_budget(&self) -> InternalGas {
+        self.gas_budget
+    }
+}
+
+/// Charge gas during a native call. If the charging fails, return early
+#[macro_export]
+macro_rules! native_charge_gas_early_exit {
+    ($native_context:ident, $gas_left:ident, $cost:expr) => {{
+        use move_core_types::vm_status::sub_status::NFE_OUT_OF_GAS;
+        match $gas_left.checked_sub($cost) {
+            Some(x) => $gas_left = x,
+            None => {
+                // Exhausted all in budget. terminate early
+                return Ok(NativeResult::err(
+                    $native_context.gas_budget(),
+                    NFE_OUT_OF_GAS,
+                ));
+            }
+        }
+    }};
+}
+
+/// Total cost of a native call so far
+#[macro_export]
+macro_rules! native_gas_total_cost {
+    ($native_context:ident, $gas_left:ident) => {{
+        // Its okay to unwrap because the budget can never be less than the gas left
+        $native_context.gas_budget().checked_sub($gas_left).unwrap()
+    }};
```

Comment:
so the model is that 
1. we grab all gas left on entry (say gas budget `100`) 
2. we call `native_charge_gas_early_exit` for operations we want to meter, and if that goes out of gas we stop. So say we call it twice with 10, we end up with `gas_left` being 80. If we pass `100` we fail right there
3. at the end of a native we call `native_gas_total_cost` to report the cost outside the function into the interpreter. And the cost is budget - whatever is left, in the example 100 - 80 = 20 which is the cost.

Is that right?

Reply:
> so the model is that
> 
> 1. we grab all gas left on entry (say gas budget `100`)
> 2. we call `native_charge_gas_early_exit` for operations we want to meter, and if that goes out of gas we stop. So say we call it twice with 10, we end up with `gas_left` being 80. If we pass `100` we fail right there
> 3. at the end of a native we call `native_gas_total_cost` to report the cost outside the function into the interpreter. And the cost is budget - whatever is left, in the example 100 - 80 = 20 which is the cost.
> 
> Is that right?

Precisely!

Reply:
> so the model is that
> 
> 1. we grab all gas left on entry (say gas budget `100`)
> 2. we call `native_charge_gas_early_exit` for operations we want to meter, and if that goes out of gas we stop. So say we call it twice with 10, we end up with `gas_left` being 80. If we pass `100` we fail right there
> 3. at the end of a native we call `native_gas_total_cost` to report the cost outside the function into the interpreter. And the cost is budget - whatever is left, in the example 100 - 80 = 20 which is the cost.
> 
> Is that right?

Now on further thought, I can the `gas_left` inside `NativeContext`, which will make things cleaner.
I can tackle this in a subsequent PR